File: Java/Hard/Accepted/contains-duplicate-iii_0.java
class Solution {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int indexDiff, int valueDiff) {
        
        TreeSet<Long> set = new TreeSet<>();
        for (int i = 0; i < nums.length; i++) {
            Long ceiling = set.ceiling((long) nums[i] - valueDiff);
            if (ceiling != null && ceiling <= (long) nums[i] + valueDiff) {
                return true;
            }
            set.add((long) nums[i]);
            if (i >= indexDiff) {
                set.remove((long) nums[i - indexDiff]);
            }
        }
        return false;
    }
}

File: Rust/Hard/Accepted/contains-duplicate-iii_0.rs
impl Solution {
    pub fn contains_nearby_almost_duplicate(nums: Vec<i32>, index_diff: i32, value_diff: i32) -> bool {
        
        let index_diff = index_diff as usize;
        let value_diff = value_diff as i64;
        let mut set = std::collections::BTreeSet::new();
        for i in 0..nums.len() {
            let n = nums[i] as i64;
            if let Some(&x) = set.range(n - value_diff..=n + value_diff).next() {
                return true;
            }
            set.insert(n);
            if i >= index_diff {
                set.remove(&(nums[i - index_diff] as i64));
            }
        }
        false
    }
}